### PR TITLE
generalize the polling of `monitorNodes` between node and master

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
This PR fixes two issues. The first commit is a problem introduced by the new load balancing feature. The second commit fixes an issue that has always been existing for manual assignment of egress IPs with opensift-sdn, but which we can/might want to fix now given the first commit.

I am not certain the second commit is needed, we could just argue that it's never really worked so no need to fix it now when this plugin is reaching a deprecation phase. I added it however as I'd like to get some input on if people think it could be useful.

/assign @danwinship @JacobTanenbaum

### CODE REVIEW

File: pkg/network/node/node.go [hunk 1]
```diff
@@ -369,7 +369,9 @@ func (node *OsdnNode) Start() error {
 		if err := node.SetupEgressNetworkPolicy(); err != nil {
 			return err
 		}
-		if err := node.egressIP.Start(node.networkInformers, node.nodeIPTables); err != nil {
+		nodeinformer := node.kubeInformers.Core().V1().Nodes()
+		nodeinformer.Informer().GetController()
```
Comment (danwinship) [inline pkg/network/node/node.go:373]: what?
(Docs for "GetController": "`// GetController is deprecated, it does nothing useful`)

also, maybe better to just pass `node.kubeInformers` to `Start` and let it pick out the specific informers it needs, rather than having node.go know

File: pkg/network/node/vxlan_monitor_test.go [hunk 2]
```diff
@@ -1,315 +0,0 @@
...
-func peekUpdate(updates chan struct{}, evm *egressVXLANMonitor) []egressVXLANNode {
-	select {
-	case <-updates:
-		return evm.GetUpdates()
-	default:
-		return nil
-	}
-}
-
-func TestEgressVXLANMonitor(t *testing.T) {
```
Comment (danwinship) [inline pkg/network/node/vxlan_monitor_test.go:42]: seems like probably we need more testing in `egressip_test.go` to make up for the lost testing in `vxlan_monitor_test.go`

File: pkg/network/common/egressip.go [hunk 3]
```diff
@@ -290,6 +303,86 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+}
+
+const (
+	DefaultPollInterval = 5 * time.Second
+	repollInterval      = time.Second
+	maxRetries          = 2
+)
+
+func (eit *EgressIPTracker) Poll() (bool, error) {
+	klog.Infof("Starting poll")
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:325]: Yeah, indeed something I forgot to clean up before submitting the patch. I'll remove these.

File: pkg/network/node/node.go (hunk 1 above)
Comment (alexanderConstantinescu) [reply pkg/network/node/node.go:373]: This is also something left over from my testing, I had some hick-ups while writing this and saw this comment: https://github.com/openshift/sdn/blob/46fd1cd3346980b0acc869e3dfdc8e6c816f7de2/pkg/network/master/master.go#L79-L84 thinking it might solve my problem. 

I'll remove that.

File: pkg/network/master/egressip.go [hunk 4]
```diff
@@ -121,11 +108,11 @@ func (eim *egressIPManager) maybeDoUpdateEgressCIDRs() (bool, error) {
 		}
 	}
 
-	eim.monitorNodes = monitorNodes
+	eim.tracker.MonitorNodes = monitorNodes
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply pkg/network/master/egressip.go:111]: I agree here, but it's not that evident. Master and node have two different set of rules for how they wish to monitor nodes. Nodes only decide to monitor when the egress IP is not assigned to the specific node in question, and master only monitors when `egressCIDRs` are set (and subsequently stops monitoring when the assigned egress IPs, calculated from `egressCIDRs`, is empty). I am not sure we can have `common` manage that by itself, and even if it would be possible, it might require more refactor than we're willing to do. What I can do to avoid master and node tampering with common's data, is: define some public methods (`SetMonitorNodes` for master, `AddEgressIP`/`RemoveEgressIP` for node) which would do what master and node already do, but inside common instead.

File: pkg/network/master/egressip.go [hunk 5]
```diff
@@ -121,11 +108,11 @@ func (eim *egressIPManager) maybeDoUpdateEgressCIDRs() (bool, error) {
...
 	if len(monitorNodes) > 0 {
 		if eim.stop == nil {
 			eim.stop = make(chan struct{})
-			go eim.poll(eim.stop)
+			go utilwait.PollUntil(common.DefaultPollInterval, eim.tracker.Poll, eim.stop)
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply pkg/network/master/egressip.go:115]: Here I am using the poll API that node (vxlan_monitor) defined. I thought it was cleaner utilizing the `utilwait` functionality as much as possible. I guess you're saying that `eim.poll` had a tighter retry loop than `evm.poll` did (0 sec vs. 1 sec), and that we should use that instead?

File: pkg/network/node/vxlan_monitor_test.go (hunk 2 above)
Comment (alexanderConstantinescu) [reply pkg/network/node/vxlan_monitor_test.go:42]: I'll try to write something up. Although, `vxlan_monitor_test.go` only tested the logic that founded itself on parsing OVS stats. There were no tests for what concerns polling nodes and mocking connectivity issues.

File: pkg/network/master/egressip.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (danwinship) [reply pkg/network/master/egressip.go:115]: no, I mean, why make the caller do the `utilwait.PollUtil` (and have to know the right `DefaultPollInterval` to use) rather than just having `eit.Poll` do that itself

File: pkg/network/node/vxlan_monitor_test.go (hunk 2 above)
Comment (danwinship) [reply pkg/network/node/vxlan_monitor_test.go:42]: ok, if we haven't lost any non-ovs-parsing testing then we don't need anything more

File: pkg/network/node/node.go (hunk 1 above)
Comment (danwinship) [reply pkg/network/node/node.go:373]: lol. yay crufty code

File: pkg/network/master/egressip.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply pkg/network/master/egressip.go:115]: OK, that concern is handled by the public methods I described here: https://github.com/openshift/sdn/pull/291#discussion_r620981098 

Given the existence of a a public `AddEgressIP` function which `EgressIPTracker` would expose, we can mask `DefaultPollInterval` from the caller and keep it internal to `common`. I will re-push soon with a version of this.

**Comment by alexanderConstantinescu** [comment]
> Sharing the code between node/ and master/ is definitely the right move. I just reviewed the two commits together. You should squash them.

Squashed. Please take another look when you get a chance.

/bugzilla cc-qa

@huiran0826 : could you verify both bugs listed in the commit and verify that this solves them? I have tested this on a local setup on my computer, but would like your input

**Comment by alexanderConstantinescu** [comment]
> @huiran0826 : could you verify both bugs listed in the commit and verify that this solves them? I have tested this on a local setup on my computer, but would like your input

Thank you for the help! As per comment: https://bugzilla.redhat.com/show_bug.cgi?id=1948436#c1 and https://bugzilla.redhat.com/show_bug.cgi?id=1948969#c1 this patch fixes both bugs.

File: pkg/network/common/egressip.go [hunk 6]
```diff
@@ -90,24 +97,27 @@ func NewEgressIPTracker(watcher EgressIPWatcher) *EgressIPTracker {
 	return &EgressIPTracker{
 		watcher: watcher,
 
-		nodes:            make(map[ktypes.UID]*nodeEgress),
-		nodesByNodeIP:    make(map[string]*nodeEgress),
+		nodes:            make(map[ktypes.UID]*NodeEgress),
+		nodesByNodeIP:    make(map[string]*NodeEgress),
```
Comment (danwinship) [inline pkg/network/common/egressip.go:101]: (It would be nice to have a separate commit for renaming/exporting NodeEgress, so that all the parts that are just capitalization are separate from the stuff that actually changes functionality)

File: pkg/network/common/egressip.go [hunk 7]
```diff
@@ -290,6 +300,130 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+func nodeIsReady(node *v1.Node) bool {
+	nodeReady := true
+	for _, cond := range node.Status.Conditions {
+		if cond.Type == v1.NodeReady {
+			if cond.Status == v1.ConditionFalse || cond.Status == v1.ConditionUnknown {
+				nodeReady = false
+			}
+		}
+	}
+	return nodeReady
```
Comment (danwinship) [inline pkg/network/common/egressip.go:312]: this function doesn't actually need `nodeReady`. You can just `return false` if you find a non-ready condition, and `return true` at the end if you don't

File: pkg/network/common/egressip.go [hunk 8]
```diff
@@ -290,6 +300,130 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+			eit.stop = make(chan struct{})
+			go utilwait.PollUntil(DefaultPollInterval, eit.poll, eit.stop)
+		}
+	} else {
+		if eit.stop != nil {
+			close(eit.stop)
+			eit.stop = nil
+		}
+	}
+}
```
Comment (danwinship) [inline pkg/network/common/egressip.go:328]: This really feels like it needs a mutex... or if not, there need to be comments explaining why it doesn't...

(applies to all of the newly-public API...)

File: pkg/network/common/egressip.go [hunk 9]
```diff
@@ -290,6 +300,130 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+		delete(eit.monitorNodes, nodeIP)
+		if len(eit.monitorNodes) == 0 && eit.stop != nil {
+			close(eit.stop)
+			eit.stop = nil
+		}
+	}
+}
+
+const (
+	DefaultPollInterval = 5 * time.Second
```
Comment (danwinship) [inline pkg/network/common/egressip.go:364]: no longer needs to be exported, so `defaultPollInterval`

File: pkg/network/common/egressip.go (hunk 7 above)
Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:312]: I agree, I will change that.

File: pkg/network/common/egressip.go (hunk 8 above)
Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:328]: It does not need a mutex, the lock on `EgressIPTracker` will already be held once this function is accessed since this function is called from `syncEgressNodeState`, which in turn is called from `syncEgressIPs` which is always called with the `EgressIPTracker` lock held. I will however add a comment mentioning this, and mention that we always expect the caller to hold the lock

Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:328]: Oh, sorry, it wasn't clear from the Github review which function your comment was referencing. I now see your comment mainly referenced `SetMonitorNodes`. This function is only called from `maybeDoUpdateEgressCIDRs` (sdn-controller) which already has its own lock held: `egressIPManager` lock. We can thus assume that all sdn-controller go-routines will access this function safely. The only concern I had when going over this in my mind was w.r.t `check`, which does not hold the `EgressIPTracker` lock, it however ends up calling `SetNodeOffline` which does...but I think it might be safer if I move the locking from `SetNodeOffline` to `check`

Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:328]: Circling back on my latest comment, moving the locking from `SetNodeOffline` to `check` will obviously mean that we lock much more often (every 5 seconds) with a possibility of locking for an extended period of time (depending on how many nodes we need to ping and the time it takes to process the answer). I'm not sure we want to do that? Maybe this is the reason we never did so to begin with?

Comment (danwinship) [reply pkg/network/common/egressip.go:328]: > Maybe this is the reason we never did so to begin with?

If the code doesn't hold a lock now then either (a) the current design of the code ensures that it doesn't need to hold a lock there, or (b) it's buggy and we just haven't gotten bitten yet. If the new design of the code doesn't ensure that it doesn't need to hold a lock there (eg, because now it depends on what the manager is doing, which is outside of the tracker's control), then we need to hold a lock, or else rearrange the code.

If `SetMonitorNodes` is being called in one thread by the manager and `check` is iterating `monitorNodes` in another thread, then we need a lock. One possibility would be to have `check` call a `getMonitorNodes` function that holds the lock and then gets the list of nodes then drops the lock. You'd need to make sure the handling of the individual EgressNodes was thread-safe after that.

Actually... it seems like `SetMonitorNodes` is currently wrong; it needs to merge the caller-provided node data in with the current data, to ensure that the private state for each node is kept

File: pkg/network/common/egressip.go (hunk 6 above)
Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:101]: Done

**Comment by danwinship** [comment]
er, I meant have a commit renaming things *first*, then have a commit changing the functionality. So that everything still compiles after both commits.

Also, please squash the fixes. It's really easier to review that way

File: pkg/network/common/egressip.go [hunk 10]
```diff
@@ -290,6 +300,153 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+	return true
+}
+
+// SetMonitorNodes updates EgressIPTracker with the node data provided and
+// starts a go-routine (if one is not already running) which polls monitorNodes
+// to check if they are ready and reachable
+func (eit *EgressIPTracker) SetMonitorNodes(monitorNodes map[string]*NodeEgress) {
+	eit.Lock()
+	defer eit.Unlock()
+	eit.monitorNodes = monitorNodes
```
Comment (danwinship) [inline pkg/network/common/egressip.go:320]: This still has the problem I was talking about before with losing information. Every time a `NetNamespace` or `HostSubnet` gets updated and the master calls `SetMonitorNodes`, it will cause the tracker to forget the `sdnIP`, `requestedIPs`, `requestedCIDRs`, `parsedCIDRs`, `activeEgressIPs`, `offline`, and `retries` of all the nodes it had been tracking before.

You have to do a more complicated sort of "sync" operation here, adding elements to and removing elements from `eit.monitorNodes` so that at the end they have the same keys, but whenever there's overlap, it needs to keep the original copy from `eit.monitorNodes`, not replace it with the one from `monitorNodes`

File: pkg/network/common/egressip.go [hunk 11]
```diff
@@ -290,6 +300,153 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+// GetMonitorNodes gets the node data from EgressIPTracker
+func (eit *EgressIPTracker) GetMonitorNodes() map[string]*NodeEgress {
+	eit.Lock()
+	defer eit.Unlock()
+	return eit.monitorNodes
+}
+
+// AddEgressIP assumes the EgressIPTracker lock is already held. The access to
+// this function call is currently percolated down from `syncEgressIPs`, which
+// always has the lock held. It starts a go-routine (if one is not already
```
Comment (danwinship) [inline pkg/network/common/egressip.go:343]: Ugh. This is pretty sketchy; an exported API shouldn't be making assumptions about its caller like that...
(I agree that the current state of the code is _correct_; it's just not very good API design on our part.)
Maybe leave a FIXME comment here pointing out that this could be made better?

**Review by danwinship — COMMENTED** [review]
Blah. This is more complicated than it had seemed originally. Maybe we should go back to your original simpler not-merging-the-master-and-node-together fix for now?

File: pkg/network/common/egressip.go (hunk 10 above)
Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:320]: > This still has the problem I was talking about before with losing information.

No, I don't think what you're describing is correct. In the most recent, and current, version of this patch I added: https://github.com/openshift/sdn/blob/0551a495b59c87b6d5da5144a76e6ba2da4a717f/pkg/network/master/egressip.go#L90-L92 which means that `maybeDoUpdateEgressCIDRs` only does `SetMonitorNodes` for nodes not currently monitored. Since the access here: https://github.com/openshift/sdn/blob/0551a495b59c87b6d5da5144a76e6ba2da4a717f/pkg/network/master/egressip.go#L82  is thread-safe, there shouldn't be any problem w.r.t overriding data that's currently existing in `monitorNodes` 

I need will test what I am saying right now and confirm that this works.

File: pkg/network/common/egressip.go (hunk 11 above)
Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:343]: I will add a FIXME

File: pkg/network/common/egressip.go (hunk 10 above)
Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:320]: Actually, I think I see what you mean now. Indeed if any item in `monitorNodes` happens to be modified in between `GetMonitorNodes` and `SetMonitorNodes` in `maybeDoUpdateEgressCIDRs`, we will override the modified field with the old one. If we however instead do 

```go
if _, exists := monitorNodes[hs.HostIP]; !exists {  	
    eim.tracker.AddToMonitorNodes(&amp;common.NodeEgress{NodeIP: hs.HostIP, NodeName: nodeName})
} 
```
and have `AddToMonitorNodes` be thread-safe while removing `SetMonitorNodes`, we should be fine, right?

Comment (alexanderConstantinescu) [reply pkg/network/common/egressip.go:326]: Doing all of this I realized that `maybeDoUpdateEgressCIDRs` was inherently wrong. There was never any situation in which we would stop monitoring nodes (even if nodes are not marked for egress IP anymore). I've fixed this now, by implementing something that resembles the current implementation.

**Comment by alexanderConstantinescu** [comment]
> Blah. This is more complicated than it had seemed originally. Maybe we should go back to your original simpler not-merging-the-master-and-node-together fix for now?

Honestly, I can't even tell anymore what the old implementation was exactly...Essentially, the code now might look a bit off for what concerns the `EgressIPTracker` API, but it works. I would rather get this PR in as it is now, fix the problem at hand (which is blocking 4.8) and figure out a way to clean this up in the future.

**Comment by alexanderConstantinescu** [comment]
Putting a "hold" on this until after code freeze, giving us time to think of a better way to do this.

/retitle  generalize the polling of `monitorNodes` between node and master

File: pkg/network/common/egressip.go [hunk 12]
```diff
@@ -290,6 +300,174 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+		activeEgressIPs: sets.NewString(egressIP),
+		NodeName:        nodeName,
+	})
+	if eit.getMonitorNodesLen() == 1 && eit.stop == nil {
+		eit.stop = make(chan struct{})
+		go utilwait.PollUntil(defaultPollInterval, eit.poll, eit.stop)
+	}
+}
+
+func (eit *EgressIPTracker) getMonitorNodesLen() int {
```
Comment (danwinship) [inline pkg/network/common/egressip.go:369]: The fact that you have to implement this by hand (in this way) suggests that a `sync.Map` isn't really the data structure we want here. Why did you change this from a mutex+map again? I don't see that in the comments.

File: pkg/network/common/egressip.go [hunk 13]
```diff
@@ -290,6 +300,174 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
...
+			}
+		}
+	}
+	return true
+}
+
+// SetMonitorNodes updates EgressIPTracker with the node data provided and
+// starts a go-routine (if one is not already running) which polls monitorNodes
+// to check if they are ready and reachable.
+func (eit *EgressIPTracker) SetMonitorNodes(monitorNodes []*NodeEgress) {
```
Comment (danwinship) [inline pkg/network/common/egressip.go:317]: So this fixes the problem of "not overwriting the tracker's data with blank objects from the caller" now, but it makes the whole thing kind of weird... We mostly aren't actually using the objects passed in from the caller, so we didn't really need to export NodeEgress as a type; it would have made more sense to just use a different type here that's just `struct { Name string; IP string }`, or just use a `map[string]string`...

**Review by danwinship — COMMENTED** [review]
This is no longer fixing any bug, right? It's just trying to consolidate code between the master and node implementations? Maybe we should just not bother doing that...

**Comment by alexanderConstantinescu** [comment]
> This is no longer fixing any bug, right? It's just trying to consolidate code between the master and node implementations? Maybe we should just not bother doing that...

It does fix bug: https://bugzilla.redhat.com/show_bug.cgi?id=1948969 (excuse me, I must have un-linked the PR and Bugzilla at some point). We could however close the BZ as it's more of usability problem than anything else. Essentially: sdn health checks egress node by checking if they are reachable (can be curled) and sdn-controller also checks if the kubelet is reporting a `Ready` state for the egress nodes. QE was expecting this to be the case for sdn too, hence this PR aligning the behavior.

**Comment by alexanderConstantinescu** [comment]
[Discussed off-band] Closing this PR, reasons are:

- Not a critical issue (if there is a customer case in the future, we could re-open it)
- Underestimated complexity of implementing it
- Not enough time to re-work the PR and fix the underestimated complexity

/close
